src/request_queue.js

```diff
@@ -240,165 +268,225 @@ export class RequestQueue {
+     * call {@link RequestQueue#reclaimRequest} instead, so that the queue will give the request to some other consumer
+     * in another call to the `fetchNextRequest` function.
+     *
+     * Note that the `null` return value doesn't mean the queue processing finished,
+     * it means there are currently no pending requests.
+     * To check whether all requests in queue were finished, use {@link RequestQueue#isFinished} instead.
      *
-     * @returns {Promise<Request>} Returns the request object, or `null` if there are no more pending requests.
+     * @returns {Promise<Request>}
+     * Returns the request object or `null` if there are no more pending requests.
      */
-    fetchNextRequest() {
-        return this
-            ._ensureHeadIsNonEmpty()
-            .then(() => {
-                const nextId = this.queueHeadDict.removeFirst();
-
-                // We are likely done at this point.
-                if (!nextId) return null;
+    async fetchNextRequest() {
+        await this._ensureHeadIsNonEmpty();
 
-                this._addToInProgress(nextId);
+        const nextRequestId = this.queueHeadDict.removeFirst();
 
-                return this
-                    .getRequest(nextId)
-                    .then((request) => {
-                        // TODO: What happens when getRequest() fails? Shouldn't we run the code below too???
-                        // We need to handle this situation because request may not be available
-                        // immediately after adding to the queue.
-                        if (!request) {
-                            this._removeFromInProgress(nextId);
-                            this.queueHeadDict.add(nextId, nextId, false);
-                        }
+        // We are likely done at this point.
+        if (!nextRequestId) return null;
 
-                        return request;
-                    });
+        // This should never happen, but...
+        if (this.inProgress[nextRequestId] || this.recentlyHandled.get(nextRequestId)) {
+            log.warning('Queue head returned a request that is already in progress?!', {
+                nextRequestId,
+                inProgress: !!this.inProgress[nextRequestId],
+                recentlyHandled: !!this.recentlyHandled.get(nextRequestId),
             });
+            return null;
+        }
+
+        this.inProgress[nextRequestId] = true;
+
+        let request;
+        try {
+            request = await this.getRequest(nextRequestId);
+        } catch (e) {
+            // On error, remove the request from in progress, otherwise it would be there forever
+            delete this.inProgress[nextRequestId];
+            throw e;
+        }
+
+        // NOTE: It can happen that the queue head index is inconsistent with the main queue table. This can happen in two situations:
+
+        // 1) Queue head index is ahead of the main table and the request is not present in the table yet.
```

Comment:
I think this is getting out of hand. We'll need to encapsulate this better. Even with the comment I'm having a hard time understanding what's going on.